InfiniTAM/infinitam_ros_node.cpp

```diff
@@ -190,113 +233,286 @@ bool InfinitamNode::startInfinitam(std_srvs::SetBool::Request& request,
                  "The mesh has too few triangles, only: %d",
                  main_engine_->GetMesh()->noTotalTriangles);
 
+  std::size_t point_number = 0;
+
   // All vertices of the mesh are stored in the pcl point cloud.
   for (int64 i = 0; i < main_engine_->GetMesh()->noTotalTriangles * 3;
-      i = i + 3) {
-    point_cloud_pcl->points[i].x = triangleArray[i].p0.x;
-    point_cloud_pcl->points[i].y = triangleArray[i].p0.y;
-    point_cloud_pcl->points[i].z = triangleArray[i].p0.z;
-
-    point_cloud_pcl->points[i + 1].x = triangleArray[i].p1.x;
-    point_cloud_pcl->points[i + 1].y = triangleArray[i].p1.y;
-    point_cloud_pcl->points[i + 1].z = triangleArray[i].p1.z;
-
-    point_cloud_pcl->points[i + 2].x = triangleArray[i].p2.x;
-    point_cloud_pcl->points[i + 2].y = triangleArray[i].p2.y;
-    point_cloud_pcl->points[i + 2].z = triangleArray[i].p2.z;
+       i = i + 3) {
+    point_cloud_pcl->points[point_number].x = triangleArray[i].p0.x;
+    point_cloud_pcl->points[point_number].y = triangleArray[i].p0.y;
+    point_cloud_pcl->points[point_number].z = triangleArray[i].p0.z;
+    ++point_number;
+
+    point_cloud_pcl->points[point_number].x = triangleArray[i].p1.x;
+    point_cloud_pcl->points[point_number].y = triangleArray[i].p1.y;
+    point_cloud_pcl->points[point_number].z = triangleArray[i].p1.z;
+    ++point_number;
+
+    point_cloud_pcl->points[point_number].x = triangleArray[i].p2.x;
+    point_cloud_pcl->points[point_number].y = triangleArray[i].p2.y;
+    point_cloud_pcl->points[point_number].z = triangleArray[i].p2.z;
+    ++point_number;
   }
-  ROS_INFO("got out_cloud");
+  LOG(INFO) << "extractITMMeshToPclCloud end.";
+}
 
-  if (rm_triangle_in_cuda_memory) {
-    delete cpu_triangles;
+bool InfinitamNode::convertPolygonMeshToRosMesh(
+    const pcl::PolygonMesh::Ptr polygon_mesh_ptr,
+    shape_msgs::Mesh::Ptr ros_mesh_ptr) {
+  pcl_msgs::PolygonMesh pcl_msg_mesh;
+
+  pcl_conversions::fromPCL(*polygon_mesh_ptr, pcl_msg_mesh);
+
+  sensor_msgs::PointCloud2Modifier pcd_modifier(pcl_msg_mesh.cloud);
+
+  size_t size = pcd_modifier.size();
+
+  ros_mesh_ptr->vertices.resize(size);
+
+  ROS_INFO_STREAM("polys: " << pcl_msg_mesh.polygons.size()
+                            << " vertices: " << pcd_modifier.size());
+
+  sensor_msgs::PointCloud2ConstIterator<float> pt_iter(pcl_msg_mesh.cloud, "x");
+
+  for (size_t i = 0; i < size; i++, ++pt_iter) {
+    ros_mesh_ptr->vertices[i].x = pt_iter[0];
+    ros_mesh_ptr->vertices[i].y = pt_iter[1];
+    ros_mesh_ptr->vertices[i].z = pt_iter[2];
+  }
+
+  ROS_INFO_STREAM("Updated vertices");
+
+  ros_mesh_ptr->triangles.resize(polygon_mesh_ptr->polygons.size());
+
+  for (size_t i = 0; i < polygon_mesh_ptr->polygons.size(); ++i) {
+    if (polygon_mesh_ptr->polygons[i].vertices.size() < 3) {
+      ROS_WARN_STREAM("Not enough points in polygon. Ignoring it.");
+      continue;
+    }
+
+    for (int j = 0; j < 3; ++j) {
```

Comment:
`j < polygon_mesh_ptr->polygons[i].vertices.size()` or is the polygon always defined as triangle?